lib/datadog/statsd/telemetry.rb

```diff
@@ -40,25 +41,54 @@ def reset
         @next_flush_time = Time.now.to_i + @flush_interval
       end
 
+      def sent(metrics: 0, events: 0, service_checks: 0, bytes: 0, packets: 0)
+        @metrics += Integer(metrics) if metrics > 0
+        @events += Integer(events) if events > 0
+        @service_checks += Integer(service_checks) if service_checks > 0
+
+        @bytes_sent += Integer(bytes) if bytes > 0
+        @packets_sent += Integer(packets) if packets > 0
+      end
+
+      def dropped(bytes: 0, packets: 0)
+        @bytes_dropped += Integer(bytes) if bytes > 0
+        @packets_dropped += Integer(packets) if packets > 0
+      end
+
       def flush?
-        if @next_flush_time < Time.now.to_i
-          return true
-        end
-        return false
+        @next_flush_time < Time.now.to_i
       end
 
       def flush
         return '' if @disabled
 
         # using shorthand syntax to reduce the garbage collection
-        return %Q(
-datadog.dogstatsd.client.metrics:#{@metrics}|#{COUNTER_TYPE}|##{@tags}
-datadog.dogstatsd.client.events:#{@events}|#{COUNTER_TYPE}|##{@tags}
-datadog.dogstatsd.client.service_checks:#{@service_checks}|#{COUNTER_TYPE}|##{@tags}
-datadog.dogstatsd.client.bytes_sent:#{@bytes_sent}|#{COUNTER_TYPE}|##{@tags}
-datadog.dogstatsd.client.bytes_dropped:#{@bytes_dropped}|#{COUNTER_TYPE}|##{@tags}
-datadog.dogstatsd.client.packets_sent:#{@packets_sent}|#{COUNTER_TYPE}|##{@tags}
-datadog.dogstatsd.client.packets_dropped:#{@packets_dropped}|#{COUNTER_TYPE}|##{@tags})
+        %Q(
+datadog.dogstatsd.client.metrics:#{@metrics}|#{COUNTER_TYPE}|##{serialized_tags}
+datadog.dogstatsd.client.events:#{@events}|#{COUNTER_TYPE}|##{serialized_tags}
+datadog.dogstatsd.client.service_checks:#{@service_checks}|#{COUNTER_TYPE}|##{serialized_tags}
+datadog.dogstatsd.client.bytes_sent:#{@bytes_sent}|#{COUNTER_TYPE}|##{serialized_tags}
+datadog.dogstatsd.client.bytes_dropped:#{@bytes_dropped}|#{COUNTER_TYPE}|##{serialized_tags}
+datadog.dogstatsd.client.packets_sent:#{@packets_sent}|#{COUNTER_TYPE}|##{serialized_tags}
+datadog.dogstatsd.client.packets_dropped:#{@packets_dropped}|#{COUNTER_TYPE}|##{serialized_tags})
+      end
+
+      private
+      attr_reader :global_tags
+      attr_reader :transport_type
+
+      def serialized_tags
+        # TODO: Karim: I don't know why but telemetry tags are serialized
+        # before global tags so by refactoring this, I am keeping the same behavior
+        @serialized_tags ||= Serialization::TagSerializer.new(telemetry_tags).format(global_tags)
+      end
+
+      def telemetry_tags
+        {
+          client: 'ruby',
+          client_version: VERSION,
+          client_transport: transport_type,
+        }
```

Comment:
The method telemetry_tags is only called once because serialized_tags is memoized, but you're right, we should build the tags in the constructor as we have everything.

Reply:
And that will probably be more clear for the reader because I've also been confused the same way Albert has been.